### ISSUE TITLE
Optimize filtering to 1 value of a grouping dimension to not need to group on that dimension

### ISSUE THREAD

**Issue by cdeszaq** [yahoo/fili#171]
When filtering down to a single value in a dimension that is also being grouped on, the grouping dimension can be removed from the query sent to Druid and instead be added to each response time bucket in post-Druid processing as an annotation.

At the most extreme, if all grouping dimensions are filtered to a single value, the query could be sent as a Time Series or Top N query instead of a Group By query, giving a _significant_ performance boost.

To accomplish this, I think the following would be needed:

1. Process the grouping dimensions / filters before query building to trim grouping dimensions where possible
2. Add ResultSetMappers to the chain that will add back any removed grouping dimensions

**Comment by archolewa** [comment]
One key question I have is where we would perform this optimization. Adding request mappers to perform the annotation suggests a `RequestHandler`, but those happen after the query has already been built.

Another option would be to do it while building the `ApiRequest` (so we end up building a timeseries ApiRequest), but then we don't have a way to add mappers without polluting the `ApiRequest`. 

Would it be an additional stage between request contstruction and table selection, or would we do it after table selection but just before building the query?

**Comment by cdeszaq** [comment]
Yeah, "where" the coercion would happen I left as an exercise for the reader 😉. We don't really have a life cycle that I can think of off the top of my head where this could happen. We'd likely need to have a new "thing" before query building that's able to hand along a ResultSetMapper somehow.

**Comment by michael-mclawhorn** [comment]
Do we have confidence in these 'significant ' boost numbers?  I've heard a lot of chatter about performance pains that I'd really like to see replicated under real world conditions. 

Sent from Yahoo Mail on Android 
 
  On Wed, Feb 15, 2017 at 1:35 PM, Rick Jensen<notifications@github.com> wrote:   
Yeah, "where" the coercion would happen I left as an exercise for the reader 😉 . We don't really have a life cycle that I can think of off the top of my head where this could happen. We'd likely need to have a new "thing" before query building that's able to hand along a ResultSetMapper somehow.

**Comment by cdeszaq** [comment]
@michael-mclawhorn Yes, we've seen at least a few queries that have a 3x performance boost (1.7s vs. 5.6s) when run with and without a grouping dimension that fit the above pattern. I don't have any hard test data other than a few queries run against a couple of implementations, but they've been fairly consistent.

As far as possible design, here are some ideas that have been tossed around offline:

- We need to be able to have a step that does query modifications (API request, some sort of logical query, or Druid queries)
- As part of the query modifications, we need a place to put, hold, and use ResultSetMappers to do the annotation
- The may be done by having a "QueryPlan" object that gets built and passed around, and possibly modified, rather than building a Query directly, and passing all of those same pieces around as individual parameters
